### PR TITLE
MatchSpec merge

### PR DESCRIPTION
**Pull request by kalefranz**
reference: https://github.com/conda/conda-build/pull/2694

### CODE REVIEW

File: tests/models/test_match_spec.py [hunk 1]
```diff
@@ -717,3 +717,36 @@ def test_parse_build_number_brackets(self):
...
+
+class MatchSpecMergeTests(TestCase):
+
+    def test_merge_version(self):
+        specs = (MatchSpec('exact'), MatchSpec('exact 1.2.3 1'), MatchSpec('exact >1.0,<2'))
+        merged_specs = MatchSpec.merge(specs)
+        print(merged_specs)
+        assert len(merged_specs) == 1
+        print(merged_specs[0])
+        assert str(merged_specs[0]) == "exact[version='>1.0,<2',build=1]"
```
Comment (msarahan) [inline tests/models/test_match_spec.py]: does this merged spec even make sense?  build numbers should only be for an exact match.  Furthermore, an exact version number (one with no relational operators and no *) should always be higher in priority than an inexact version number.

This test loses the two different kinds of behavior:
1. exact has priority over inexact
2. within inexact, there's an order that results in the tightest constraint.  Max of lower bounds, min of upper bounds.

Comment (kalefranz) [reply tests/models/test_match_spec.py]: > build numbers

I'm questioning whether that third position in the conda-build style match spec string is build or build number.  I guess it has to be build.  In which case, yes, exact match only.  In conda 4.4, build numbers support comparison operators (see #6880 and #6881).

> does this merged spec even make sense?

I think you're right.  This was just an initial commit to start chatting with @mcg1969 about this.  Very early WIP!

Comment (kalefranz) [reply tests/models/test_match_spec.py]: @msarahan Small tweak in the last commit.  This is how @mcg1969 suggested combining these.  Any thoughts?

File: tests/models/test_match_spec.py [hunk 2]
```diff
@@ -717,3 +717,60 @@ def test_parse_build_number_brackets(self):
...
+            'bounded >=1.0,<2.0', 'bounded >=1.5', 'bounded <=1.8',
+        ))
+        merged_specs = MatchSpec.merge(specs)
+        print(merged_specs)
+        assert len(merged_specs) == 2
+
+        exact_spec = next(s for s in merged_specs if s.name == 'exact')
+        bounded_spec = next(s for s in merged_specs if s.name == 'bounded')
+
+        assert str(exact_spec) == "exact[version='1.2.3,>1.0,<2',build=1]"
```
Comment (msarahan) [inline tests/models/test_match_spec.py]: This is fine.  It's more constraints than it needs to be.  My code cut them out, but I expect this is safer than my code (i.e. less likely to lose the finer details of when < and <= are not equivalent), and less prone to logic errors.  It's a little confusing looking at it, because so much of the information there is superfluous - you have to think about what bound or pin will take effect, rather than just seeing it.  That said, as long as the performance impact of having more constraints will scale well, I'm happy enough with how this is.

File: tests/models/test_match_spec.py [hunk 3]
```diff
@@ -717,3 +717,60 @@ def test_parse_build_number_brackets(self):
         #     "build_number": '>=3',
         # }
 
+
+class MatchSpecMergeTests(TestCase):
+
+    def test_merge_single_name(self):
+        specs = (MatchSpec('exact'), MatchSpec('exact 1.2.3 1'), MatchSpec('exact >1.0,<2'))
```
Comment (mbargull) [inline tests/models/test_match_spec.py]: Can you also add
```python
specs = (MatchSpec('exact 1.2.3 1'), MatchSpec('exact 1.2.3 2'))
with pytest.raises(ValueError):
    MatchSpec.merge(specs)
```
so that we have an "expected failure" test?

File: conda/models/match_spec.py [hunk 4]
```diff
@@ -404,6 +406,36 @@ def fn(self):
         assert val
         return val
 
+    @classmethod
+    def merge(cls, match_specs):
+        match_specs = tuple(cls(s) for s in match_specs)
+        grouped = groupby(lambda spec: spec.get_exact_value('name'), match_specs)
+        specs_map = {
+            name: reduce(lambda x, y: x._merge(y), specs) if len(specs) > 1 else specs[0]
```
Comment (mbargull) [inline conda/models/match_spec.py]:
```python
name: reduce(lambda x, y: x._merge(y), specs)
```
should suffice as `reduce` on a one-element sequence should return `specs[0]`.

File: conda/models/match_spec.py [hunk 5]
```diff
@@ -404,6 +406,36 @@ def fn(self):
...
+            this_component = self._match_components.get(component_name)
+            that_component = other._match_components.get(component_name)
+            if this_component is None and that_component is None:
+                continue
+            elif this_component is None:
+                final_components[component_name] = that_component
+            elif that_component is None:
+                final_components[component_name] = this_component
+            else:
+                final_components[component_name] = this_component.merge(that_component)
```
Comment (mbargull) [inline conda/models/match_spec.py]: You could simply this as
```python
final_components = {}
for name, component in iteritems(self.match_components):
    if component is None:
        continue
    final_components[name] = component
for name, component in iteritems(other.match_components):
    if component is None:
        continue
    if name in final_components:
        final_components[name].merge(component)
    else:
        final_components[name] = component
```
or
```python
final_components = {name: comp for name, comp in iteritems(self._match_components) if comp is not None}
for name, component in ((name: comp) for name, comp in iteritems(other._match_components) if comp is not None):
    if name in final_components:
        final_components[name].merge(component)
    else:
        final_components[name] = component
```

File: conda/models/match_spec.py [hunk 6]
```diff
@@ -404,6 +406,36 @@ def fn(self):
         assert val
         return val
 
+    @classmethod
+    def merge(cls, match_specs):
+        match_specs = tuple(cls(s) for s in match_specs)
+        grouped = groupby(lambda spec: spec.get_exact_value('name'), match_specs)
```
Comment (mbargull) [inline conda/models/match_spec.py]: Didn't check myself, but is the `match_specs` input always sorted by `spec.get_exact_value('name')`? `groupby` groups by consecutive keys only.

Comment (mbargull) [reply conda/models/match_spec.py]: Ah, disregard that comment, you used `(cy)toolz.itertoolz.groupby` not `itertools.groupby`, so everything is fine!

File: tests/models/test_match_spec.py [hunk 7]
```diff
@@ -717,3 +717,60 @@ def test_parse_build_number_brackets(self):
...
+            'name': 'exact',
+            'version': '1.2.2',
+            'build': '1',
+            'build_number': 1,
+        })
+
+    def test_merge_multiple_name(self):
+        specs = tuple(MatchSpec(s) for s in (
+            'exact', 'exact 1.2.3 1', 'exact >1.0,<2',
+            'bounded >=1.0,<2.0', 'bounded >=1.5', 'bounded <=1.8',
```
Comment (mbargull) [inline tests/models/test_match_spec.py]: How about intermixing the specs to make sure order doesn't matter? E.g.:
```python
            'exact', 'exact 1.2.3 1',
            'bounded >=1.0,<2.0', 'bounded >=1.5', 'bounded <=1.8',
            'exact >1.0,<2',
```

File: conda/models/match_spec.py (hunk 5 above)
Comment (kalefranz) [reply conda/models/match_spec.py]: I plugged both of these in.  They both have four test failures with the additional tests I added after the most recent commit.

File: tests/models/test_match_spec.py (hunk 7 above)
Comment (kalefranz) [reply tests/models/test_match_spec.py]: done

File: tests/models/test_match_spec.py (hunk 3 above)
Comment (kalefranz) [reply tests/models/test_match_spec.py]: Done.  Plus a bunch more tests added.

File: conda/models/channel.py [hunk 8]
```diff
@@ -368,7 +368,7 @@ def _get_channel_for_name_helper(name):
         # stripping off path threw information away from channel_name (i.e. any potential subname)
         # channel.name *should still be* channel_name
         channel = copy(channel)
-        channel.name = channel_name
+        channel.name = _stripped
```
Comment (kalefranz) [inline conda/models/channel.py]: This one-line fix could be a separate PR. But oh well. It's at least in its own commit.  The commit has a test, which is included in `tests/models/test_channel.py`.

File: tests/models/test_match_spec.py (hunk 2 above)
Comment (kalefranz) [reply tests/models/test_match_spec.py]: Yeah this is what @mcg1969 suggested, exactly because of the logical complexity of reducing to a minimal form.  I think this is the solution we should go with for now, with the goal of implementing the minimal-form reduction in the future.

File: conda/models/match_spec.py (hunk 5 above)
Comment (mbargull) [reply conda/models/match_spec.py]: Writing code directly in GitHub is a **bad** habit...
Sorry, forgot the assignment in
```python
final_components[name] = final_components[name].merge(component)
```
Edited the comment above with that fix.

**Comment by msarahan** [comment]
Merging to facilitate testing for 4.5.0rc release.  Thanks for your comments @mbargull - I hope we'll keep addressing them in the release.
